SpatialGDK/Source/SpatialGDK/Private/Interop/InitialOnlyFilter.cpp

```diff
@@ -0,0 +1,135 @@
+#include "Interop/SpatialOSDispatcherInterface.h"
+#include "Interop/SpatialReceiver.h"
+
+DEFINE_LOG_CATEGORY(LogInitialOnlyFilter);
+
+namespace SpatialGDK
+{
+InitialOnlyFilter::InitialOnlyFilter(USpatialNetDriver* InNetDriver)
+	: NetDriver(InNetDriver)
+{
+}
+
+bool InitialOnlyFilter::HasInitialOnlyData(Worker_EntityId EntityId) const
+{
+	return (RetrievedInitialOnlyData.Find(EntityId) != nullptr);
+}
+
+bool InitialOnlyFilter::HasInitialOnlyDataOrRequest(Worker_EntityId EntityId)
+{
+	if (HasInitialOnlyData(EntityId))
+	{
+		return true;
+	}
+
+	if (InflightInitialOnlyEntities.Find(EntityId) != nullptr)
+	{
+		return false;
+	}
+
+	PendingInitialOnlyEntities.Add(EntityId);
+	return false;
+}
+
+void InitialOnlyFilter::FlushRequests()
+{
+	if (PendingInitialOnlyEntities.Num() == 0)
+	{
+		return;
+	}
+
+	TArray<Worker_Constraint> EntityConstraintArray;
+	TSet<Worker_EntityId_Key> EntitiesRequested(MoveTemp(PendingInitialOnlyEntities));
+
+	for (auto EntityId : EntitiesRequested)
+	{
+		UE_LOG(LogInitialOnlyFilter, Verbose, TEXT("Requested initial only data for entity %lld."), EntityId);
+
+		Worker_Constraint Constraints{};
+		Constraints.constraint_type = WORKER_CONSTRAINT_TYPE_ENTITY_ID;
+		Constraints.constraint.entity_id_constraint.entity_id = EntityId;
+
+		EntityConstraintArray.Add(Constraints);
+
+		InflightInitialOnlyEntities.Add(EntityId);
+	}
+
+	Worker_EntityQuery InitialOnlyQuery{};
+
+	InitialOnlyQuery.constraint.constraint_type = WORKER_CONSTRAINT_TYPE_OR;
+	InitialOnlyQuery.constraint.constraint.or_constraint.constraint_count = EntityConstraintArray.Num();
+	InitialOnlyQuery.constraint.constraint.or_constraint.constraints = EntityConstraintArray.GetData();
+	InitialOnlyQuery.snapshot_result_type_component_set_id_count = 1;
+	InitialOnlyQuery.snapshot_result_type_component_set_ids = &SpatialConstants::INITIAL_ONLY_COMPONENT_SET_ID;
+
+	const Worker_RequestId RequestID = NetDriver->Connection->SendEntityQueryRequest(&InitialOnlyQuery, SpatialGDK::RETRY_UNTIL_COMPLETE);
+	EntityQueryDelegate InitialOnlyQueryDelegate;
+	InitialOnlyQueryDelegate.BindRaw(this, &InitialOnlyFilter::HandleInitialOnlyResponse);
+
+	NetDriver->Receiver->AddEntityQueryDelegate(RequestID, InitialOnlyQueryDelegate);
+
+	InflightInitialOnlyRequests.Add(RequestID, { MoveTemp(EntitiesRequested) });
+}
+
+void InitialOnlyFilter::HandleInitialOnlyResponse(const Worker_EntityQueryResponseOp& Op)
+{
+	ClearRequest(Op.request_id);
+
+	if (Op.status_code != WORKER_STATUS_CODE_SUCCESS)
+	{
+		UE_LOG(LogInitialOnlyFilter, Error, TEXT("Failed to retrieve initial only data. Code: %d, %s"), Op.status_code,
+			   UTF8_TO_TCHAR(Op.message));
+		return;
+	}
+
+	for (uint32_t i = 0; i < Op.result_count; ++i)
+	{
+		const Worker_Entity* Entity = &Op.results[i];
+		const Worker_EntityId EntityId = Entity->entity_id;
+
+		// Ensure the entity we queried is still in view.
+		if (NetDriver->Connection->GetView().Find(EntityId) != nullptr)
```

Comment:
Nit, can flip the if here to return early although I don't know if we want verbose logs in the case where this condition fails.

Reply:
The query could contain multiple entity queries, so we need to process them all. Unless I'm not following what you mean,

Reply:
I just meant we can reduce one level of nesting by putting a continue in this block. And that in that case we might also want to verbose log that we've received data for something which has now gone out of view. However I'm not fussed about the log as I'm not sure where I'd care about that.

Reply:
Can make this change 👍